File: 67.二进制求和.java
/*
 * @lc app=leetcode.cn id=67 lang=java
 *
 * [67] 二进制求和
 */

// @lc code=start
//我的题解
class Solution {
    public String addBinary(String a, String b) {
        // i j 指向字符串的最后一位
        int i = a.length() - 1, j = b.length() - 1;
        //ti t2 用于表示a b 对应当前位置
        int t1, t2;

        String c = "";
        //记录是否进位
        int flag = 0;
        while (i >= 0 || j >= 0) {

            /*对t1 t2进行初始化*/
            if (i < 0) t1 = 0;
            else t1 = a.charAt(i) - '0';
            if (j < 0) t2 = 0;
            else t2 = b.charAt(j) - '0';

            c = (t1 + t2 + flag) % 2 + c;
            if (t1 + t2 + flag >= 2) flag = 1;
            else flag=0;

            i--;
            j--;
        }
        if (flag == 1) c = "1" + c;
        return c;
    }
}
// //官方题解
// class Solution {
//     public String addBinary(String a, String b) {
//         //StringBuffer·
//         StringBuffer ans = new StringBuffer();

//         int n = Math.max(a.length(), b.length()), carry = 0;
//         for (int i = 0; i < n; ++i) {
//             carry += i < a.length() ? (a.charAt(a.length() - 1 - i) - '0') : 0;
//             carry += i < b.length() ? (b.charAt(b.length() - 1 - i) - '0') : 0;
//             ans.append((char) (carry % 2 + '0'));
//             carry /= 2;
//         }

//         if (carry > 0) {
//             ans.append('1');
//         }
//         ans.reverse();

//         return ans.toString();
//     }
// }


// @lc code=end

File: 67.二进制求和.c
/*
 * @lc app=leetcode.cn id=67 lang=c
 *
 * [67] 二进制求和
 */

// @lc code=start

void reserve(char *s)
{
    int len = strlen(s);
    for (int i = 0; i < len / 2; i++)
    {
        char t = s[i];
        s[i] = s[len - i - 1], s[len - i - 1] = t;
    }
}

char *addBinary(char *a, char *b)
{
    reserve(a);
    reserve(b);

    int len_a = strlen(a), len_b = strlen(b);
    int n = fmax(len_a, len_b), carry = 0, len = 0;
    char *ans = (char *)malloc(sizeof(char) * (n + 2));
    for (int i = 0; i < n; ++i)
    {
        carry += i < len_a ? (a[i] == '1') : 0;
        carry += i < len_b ? (b[i] == '1') : 0;
        ans[len++] = carry % 2 + '0';
        carry /= 2;
    }

    if (carry)
    {
        ans[len++] = '1';
    }
    ans[len] = '\0';
    reserve(ans);

    return ans;
}

char *addBinary(char *a, char *b) {
	int len_a = strlen(a), len_b = strlen(b);
	int n = len_a > len_b ? len_a : len_b;

	char *c = (char *)malloc(sizeof(char) * (n + 2));
	c[n] = '\0';

	n--;
	len_a--;
	len_b--;
	int flag = 0, t1, t2;
	while (len_a >= 0 || len_b >= 0) {
		if (len_a >= 0)
			t1 = a[len_a] - '0';
		else
			t1 = 0;

		if (len_b >= 0)
			t2 = b[len_b] - '0';
		else
			t2 = 0;

		c[n] = (t1 + t2 + flag) % 2 + '0';

		if ((t1 + t2 + flag) >= 2)
			flag = 1;
		else
			flag = 0;

		n--;
		len_a--;
		len_b--;
	}

	char *ssss;

	if (flag) {
		ssss = (char *)malloc(sizeof(char) * 2);
		ssss[0]= '1';
		ssss[1] = '\0';
		strcat(ssss, c);
	}

	return ssss;
}
// @lc code=end